add gloo support for reduce on GPU

janewangfb opened the pull request:
Summary: as titled

Differential Revision: D13222907

pietern commented:
The distributed tests appear to be passing in the test runs that are failing. This is probably related to the revert that happened in 5f07b33857cea7f68067d67bd3e0c77dfc09592d. Could you rebase to make sure that commit is included and repush? This is too many failures to assume it's all good.

pietern reviewed (approved):
Great, thank you!

janewangfb committed:
add gloo support for reduce on GPU (#14443)

Summary:
as titled
Pull Request resolved: https://github.com/pytorch/pytorch/pull/14443

Reviewed By: pietern

Differential Revision: D13222907

Pulled By: pietern

fbshipit-source-id: 6b1ed54bd0a21f1f1d1ac1a8edbe4ba60b44813a
